src/Orleans/IDs/GuidId.cs

```diff
@@ -0,0 +1,124 @@
+/*
+Project Orleans Cloud Service SDK ver. 1.0
+ 
+Copyright (c) Microsoft Corporation
+ 
+All rights reserved.
+ 
+MIT License
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
+associated documentation files (the ""Software""), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+﻿using System;
+using Orleans.Concurrency;
+using Orleans.Serialization;
+
+namespace Orleans.Runtime
+{
+    /// <summary>
+    /// Wrapper object around Guid.
+    /// Can be used in places where Guid is optional and in those cases it can be set to null and will not use the storage of an empty Guid struct.
```

Comment:
I read a bit about Nullable types. It is a struct wrapper around the original type. So I suspect it will have the same problem I wanted to solve in the first place: it will have the extra 16 bytes at least (plus the extra space for the Nullable struct fields, if any) always. And I want zero extra storage if it is null. 
The reason this saving is important is since this is Grain Reference, the most fundamental and commonly used primitive in Orleans. We want it to be as closely efficiency-wise to a simple in memory reference as possible.